### PR TITLE
ETCD-660: Add `etcd-backup-server` backup vars

### PR DESCRIPTION
**Pull request by Elbehery**
This PR adds `etcd-backup-server` backupVar.

This is part of https://issues.redhat.com/browse/ETCD-636

cc @openshift/openshift-team-etcd

### CODE REVIEW

File: pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go [hunk 1]
```diff
@@ -8,25 +8,27 @@ import (
...
-	"github.com/openshift/cluster-etcd-operator/pkg/operator/ceohelpers"
-	"github.com/openshift/cluster-etcd-operator/pkg/operator/operatorclient"
-	u "github.com/openshift/cluster-etcd-operator/pkg/testutils"
+	"go.etcd.io/etcd/api/v3/etcdserverpb"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 )
 
 func TestTargetConfigController(t *testing.T) {
```
Comment (tjungblu) [inline pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go:34]: can you add a test here please?

File: pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go [hunk 2]
```diff
@@ -155,11 +181,15 @@ func TestTargetConfigController(t *testing.T) {
 				operatorClient:        fakeOperatorClient,
 				kubeClient:            fakeKubeClient,
 				envVarGetter:          envVar,
-				backupVarGetter:       backuphelpers.NewDisabledBackupConfig(),
+				backupVarGetter:       backupVar,
 				enqueueFn:             func() {},
 				quorumChecker:         quorumChecker,
 			}
 
+			if scenario.etcdBackupSpec != nil {
```
Comment (Elbehery) [inline pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go:189]: @tjungblu  I added this scenario 

Since we dont error upon `BackupVar` I dont know what to assert upon 🤔

Comment (tjungblu) [reply pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go:189]: the output of the static pod, not sure why this was never added here

**Comment by Elbehery** [comment]
this [commit](https://github.com/openshift/cluster-etcd-operator/pull/1305/commits/ec00273d994953ae140158c9fb4ca067879ac927) contains the updated etcdpod.yaml manifest

this is just to pass the test .. it can be skipped if conflicts with other PRs

File: pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go [hunk 3]
```diff
@@ -88,14 +96,45 @@ func TestTargetConfigController(t *testing.T) {
 				u.FakeEtcdMemberWithoutServer(0),
 				u.FakeEtcdMemberWithoutServer(2),
 			},
+			etcdBackupSpec:    nil,
 			etcdMembersEnvVar: "1,3",
 		},
+		{
+			name: "Backup Var Test",
```
Comment (Elbehery) [inline pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go:103]: this test case fails, since the static pod manifest does not contains the sidecar, not the backup vars, it depends on 

https://github.com/openshift/cluster-etcd-operator/pull/1304

**Comment by Elbehery** [comment]
The failing unit tests are expected, the test case needs to assert on the update static pod manifest, but it is not merged yet https://github.com/openshift/cluster-etcd-operator/pull/1304

**Comment by Elbehery** [comment]
@Elbehery 

- [ ] Add  `${COMPUTED_BACKUP_VARS}` to this PR after https://github.com/openshift/cluster-etcd-operator/pull/1304v merges

**Comment by Elbehery** [comment]
The failure here depends on https://github.com/openshift/cluster-etcd-operator/pull/1304 being merged 

Basically the test case `TestTargetConfigController/Backup_Var_Test` needs `COMPUTED_BACKUP_VAR` for assertion. This var will be added to the manifest here, once https://github.com/openshift/cluster-etcd-operator/pull/1304 is merged 

cc @openshift/openshift-team-etcd

File: pkg/backuphelpers/backupvars.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+	if reflect.DeepEqual(b.spec, spec) {
+		return
+	}
+
+	if isEmptyEtcdBackupSpec(spec) {
+		b.enabled = false
+	} else {
+		b.enabled = true
+	}
```
Comment (tjungblu) [inline pkg/backuphelpers/backupvars.go:49]:
```suggestion
   b.enabled = (spec == nil)
```

File: pkg/backuphelpers/backupvars.go [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+type BackupVar interface {
+	AddListener(listener Enqueueable)
+	SetBackupSpec(spec backupv1alpha1.EtcdBackupSpec)
+	ArgString() string
+}
+
+type BackupConfig struct {
+	enabled   bool
+	spec      backupv1alpha1.EtcdBackupSpec
```
Comment (tjungblu) [inline pkg/backuphelpers/backupvars.go:25]:
```suggestion
	spec      *backupv1alpha1.EtcdBackupSpec
```

let's make this a pointer instead

File: pkg/backuphelpers/backupvars.go [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+	prune "github.com/openshift/cluster-etcd-operator/pkg/cmd/prune-backups"
+)
+
+type Enqueueable interface {
+	Enqueue()
+}
+
+type BackupVar interface {
+	AddListener(listener Enqueueable)
+	SetBackupSpec(spec backupv1alpha1.EtcdBackupSpec)
```
Comment (tjungblu) [inline pkg/backuphelpers/backupvars.go:19]:
```suggestion
	SetBackupSpec(spec *backupv1alpha1.EtcdBackupSpec)
```

File: pkg/cmd/prune-backups/prune.go [hunk 7]
```diff
@@ -33,10 +33,14 @@ type PruneOpts struct {
 	RetentionType      string
 	MaxNumberOfBackups int
 	MaxSizeOfBackupsGb int
+	BackupPath         string
```
Comment (tjungblu) [inline pkg/cmd/prune-backups/prune.go:36]: please remove all prune related changes from this PR

File: pkg/operator/periodicbackupcontroller/periodicbackupcontroller.go [hunk 8]
```diff
@@ -80,6 +89,16 @@ func (c *PeriodicBackupController) sync(ctx context.Context, _ factory.SyncConte
 		return fmt.Errorf("PeriodicBackupController could not list backup CRDs, error was: %w", err)
 	}
 
+	if len(backups.Items) == 0 && !c.isDefaultBackup {
```
Comment (tjungblu) [inline pkg/operator/periodicbackupcontroller/periodicbackupcontroller.go:92]: can we go back to how we initially have this set in the loop?

File: pkg/operator/periodicbackupcontroller/periodicbackupcontroller.go [hunk 9]
```diff
@@ -80,6 +89,16 @@ func (c *PeriodicBackupController) sync(ctx context.Context, _ factory.SyncConte
...
+		c.backupVarGetter.SetBackupSpec(createDefaultEtcdBackupSpec())
+		c.isDefaultBackup = true
+	}
+
+	if len(backups.Items) > 0 && c.isDefaultBackup {
+		c.backupVarGetter.SetBackupSpec(backupv1alpha1.EtcdBackupSpec{})
+		c.isDefaultBackup = false
+	}
+
 	for _, item := range backups.Items {
```
Comment (tjungblu) [inline pkg/operator/periodicbackupcontroller/periodicbackupcontroller.go:102]:
```suggestion
	for _, item := range backups.Items {
	     if item.Name == "default" {
  	        c.backupVarGetter.SetBackupSpec(&item) // guess this needs a copy as well
	     }
```

File: pkg/operator/periodicbackupcontroller/periodicbackupcontroller.go (hunk 8 above)
Comment (Elbehery) [reply pkg/operator/periodicbackupcontroller/periodicbackupcontroller.go:92]: so the reason I have used this approach, is to avoid having to install a CR for the NoConfig 

Since this feature introduces NoConfig backups, the user should not be required to apply any CR 

please correct me if I am wrong

File: pkg/cmd/prune-backups/prune.go (hunk 7 above)
Comment (Elbehery) [reply pkg/cmd/prune-backups/prune.go:36]: done 👍🏽

File: pkg/backuphelpers/backupvars.go (hunk 4 above)
Comment (Elbehery) [reply pkg/backuphelpers/backupvars.go:49]: done 👍🏽

File: pkg/backuphelpers/backupvars.go (hunk 5 above)
Comment (Elbehery) [reply pkg/backuphelpers/backupvars.go:25]: done 👍🏽

File: pkg/backuphelpers/backupvars.go (hunk 6 above)
Comment (Elbehery) [reply pkg/backuphelpers/backupvars.go:19]: done 👍🏽

File: pkg/operator/periodicbackupcontroller/periodicbackupcontroller.go (hunk 9 above)
Comment (Elbehery) [reply pkg/operator/periodicbackupcontroller/periodicbackupcontroller.go:92]: done 👍🏽

File: pkg/backuphelpers/backupvars.go [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+	b.mux.Lock()
+	defer b.mux.Unlock()
+
+	if reflect.DeepEqual(b.spec, spec) {
+		return
+	}
+
+	b.enabled = spec != nil
+
+	b.spec = spec
```
Comment (tjungblu) [inline pkg/backuphelpers/backupvars.go:47]: don't we have to make a copy here?

File: pkg/cmd/prune-backups/prune.go [hunk 11]
```diff
@@ -36,7 +36,9 @@ type PruneOpts struct {
 }
 
 func NewPruneCommand() *cobra.Command {
-	opts := PruneOpts{RetentionType: "None"}
+	opts := PruneOpts{
```
Comment (tjungblu) [inline pkg/cmd/prune-backups/prune.go:39]: unrelated to the purpose of this PR

Comment (Elbehery) [reply pkg/cmd/prune-backups/prune.go:39]: sure 👍🏽

File: pkg/cmd/backuprestore/backupserver.go [hunk 12]
```diff
@@ -39,14 +37,19 @@ func NewBackupServer(ctx context.Context) *cobra.Command {
...
 }
 
-func (b *backupServer) AddFlags(fs *pflag.FlagSet) {
+func (b *backupServer) AddFlags(cmd *cobra.Command) {
+	fs := cmd.Flags()
 	fs.BoolVar(&b.enabled, "enabled", false, "enable backup server")
 	fs.StringVar(&b.schedule, "schedule", "", "schedule specifies the cron schedule to run the backup")
 	fs.StringVar(&b.timeZone, "timezone", "", "timezone specifies the timezone of the cron schedule to run the backup")
+
+	cobra.MarkFlagRequired(fs, "enabled")
```
[the remark this answers is not in the document]
Comment (Elbehery) [reply pkg/cmd/backuprestore/backupserver.go:50]: sure 👍🏽

File: pkg/backuphelpers/backupvars.go (hunk 10 above)
Comment (Elbehery) [reply pkg/backuphelpers/backupvars.go:47]: so I make a copy in the periodicController, and pass a pointer to the copy here

File: pkg/operator/periodicbackupcontroller/periodicbackupcontroller.go [hunk 13]
```diff
@@ -260,3 +274,16 @@ func newCronJob() (*batchv1.CronJob, error) {
...
+		Schedule: "*/5 * * * *",
+		TimeZone: "UTC",
+		RetentionPolicy: backupv1alpha1.RetentionPolicy{
+			RetentionType: prune.RetentionTypeNumber,
+			RetentionNumber: &backupv1alpha1.RetentionNumberConfig{
+				MaxNumberOfBackups: 3,
+			},
+		},
+	}
+}
```
Comment (tjungblu) [inline pkg/operator/periodicbackupcontroller/periodicbackupcontroller.go:289]:
```suggestion

```

unrelated

File: pkg/operator/periodicbackupcontroller/periodicbackupcontroller.go [hunk 14]
```diff
@@ -9,33 +9,39 @@ import (
...
+)
 
 type PeriodicBackupController struct {
 	operatorClient        v1helpers.OperatorClient
 	backupsClient         backupv1client.BackupsGetter
 	kubeClient            kubernetes.Interface
 	operatorImagePullSpec string
+	backupVarGetter       backuphelpers.BackupVar
 	featureGateAccessor   featuregates.FeatureGateAccess
+	isDefaultBackup       bool
```
Comment (tjungblu) [inline pkg/operator/periodicbackupcontroller/periodicbackupcontroller.go:44]:
```suggestion
```

unrelated

File: pkg/backuphelpers/backupvars.go (hunk 10 above)
Comment (Elbehery) [reply pkg/backuphelpers/backupvars.go:47]: this way 

```
if item.Name == "default" {
			specClone := item
			c.backupVarGetter.SetBackupSpec(&specClone.Spec.EtcdBackupSpec)
		}
```

File: pkg/operator/periodicbackupcontroller/periodicbackupcontroller.go [hunk 15]
```diff
@@ -81,6 +90,11 @@ func (c *PeriodicBackupController) sync(ctx context.Context, _ factory.SyncConte
 	}
 
 	for _, item := range backups.Items {
+		if item.Name == "default" {
+			specClone := item
+			c.backupVarGetter.SetBackupSpec(&specClone.Spec.EtcdBackupSpec)
+		}
```
Comment (tjungblu) [inline pkg/operator/periodicbackupcontroller/periodicbackupcontroller.go:96]:
```suggestion
                  continue
		}
```
we don't want to reconcile it

Comment (Elbehery) [reply pkg/operator/periodicbackupcontroller/periodicbackupcontroller.go:96]: i thought i added it, my bad :'(

File: pkg/backuphelpers/backupvars.go [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+func (b *BackupConfig) ArgString() string {
+	b.mux.Lock()
+	defer b.mux.Unlock()
+
+	args := []string{"    args:"}
+	args = append(args, fmt.Sprintf("- --%s=%v", "enabled", b.enabled))
+
+	if b.spec == nil {
+		return strings.Join(args, "\n    ")
+	}
```
Comment (tjungblu) [inline pkg/backuphelpers/backupvars.go:62]:
```suggestion
	args = append(args, fmt.Sprintf("- --%s=%v", "enabled", b.enabled))

	if !b.enabled {
		return strings.Join(args, "\n    ")
	}
```

if you already create an extra variable, make use of it :)

File: pkg/backuphelpers/backupvars_test.go [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+			"enabled",
+			createEtcdBackupSpec(timezone, schedule),
+			true,
+			"    args:\n    - --enabled=true\n    - --timezone=GMT\n    - --schedule=0 */2 * * *",
+		},
+		{
+			"disabled",
+			createEtcdBackupSpec(timezone, schedule),
+			false,
+			"    args:\n    - --enabled=true\n    - --timezone=GMT\n    - --schedule=0 */2 * * *",
```
Comment (tjungblu) [inline pkg/backuphelpers/backupvars_test.go:34]: this is confusing, you say enabled=false and it ends up as `--enabled=true`?

File: pkg/backuphelpers/backupvars_test.go [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+		},
+		{
+			"empty schedule disabled",
+			createEtcdBackupSpec(timezone, ""),
+			false,
+			"    args:\n    - --enabled=true\n    - --timezone=GMT",
```
Comment (tjungblu) [inline pkg/backuphelpers/backupvars_test.go:40]: here too, disabled but `--enabled=true`

File: pkg/backuphelpers/backupvars_test.go (hunk 17 above)
Comment (Elbehery) [reply pkg/backuphelpers/backupvars_test.go:34]: so the initial state is `false`, but it will be enabled

File: pkg/backuphelpers/backupvars_test.go (hunk 18 above)
Comment (Elbehery) [reply pkg/backuphelpers/backupvars_test.go:40]: same

File: pkg/backuphelpers/backupvars.go (hunk 16 above)
Comment (Elbehery) [reply pkg/backuphelpers/backupvars.go:62]: +1

File: pkg/cmd/prune-backups/prune.go (hunk 11 above)
Comment (Elbehery) [reply pkg/cmd/prune-backups/prune.go:39]: this was already here, it is just the formatting

File: pkg/operator/periodicbackupcontroller/periodicbackupcontroller.go [hunk 19]
```diff
@@ -81,6 +87,12 @@ func (c *PeriodicBackupController) sync(ctx context.Context, _ factory.SyncConte
 	}
 
 	for _, item := range backups.Items {
+		if item.Name == defaultBackupCRName {
+			specClone := item
```
Comment (tjungblu) [inline pkg/operator/periodicbackupcontroller/periodicbackupcontroller.go:91]:
```suggestion  
                        defaultFound = true
			specClone := item
```

File: pkg/operator/periodicbackupcontroller/periodicbackupcontroller.go [hunk 20]
```diff
@@ -81,6 +87,12 @@ func (c *PeriodicBackupController) sync(ctx context.Context, _ factory.SyncConte
 	}
 
 	for _, item := range backups.Items {
```
Comment (tjungblu) [inline pkg/operator/periodicbackupcontroller/periodicbackupcontroller.go:89]: sorry I totally forgot, we need a case for when the default was removed
```suggestion
        defaultFound := false
	for _, item := range backups.Items {
```

Comment (tjungblu) [reply pkg/operator/periodicbackupcontroller/periodicbackupcontroller.go:89]: and below the loop (can't suggest) you need to set it to nil:

```
if !defaultFound {
   c.backupVarGetter.SetBackupSpec(nil)
}
```

File: pkg/backuphelpers/backupvars.go (hunk 10 above)
Comment (tjungblu) [reply pkg/backuphelpers/backupvars.go:47]: yes, but should every caller of SetBackupSpec know that they need to make a copy before?

File: pkg/backuphelpers/backupvars_test.go (hunk 17 above)
Comment (tjungblu) [reply pkg/backuphelpers/backupvars_test.go:34]: then we need to assert the actual output, not the initial state - otherwise the test makes no sense?

File: pkg/cmd/backuprestore/backupserver.go [hunk 21]
```diff
@@ -39,14 +37,17 @@ func NewBackupServer(ctx context.Context) *cobra.Command {
 			}
 		},
 	}
-	backupSrv.AddFlags(cmd.Flags())
+	backupSrv.AddFlags(cmd)
 	return cmd
 }
 
-func (b *backupServer) AddFlags(fs *pflag.FlagSet) {
+func (b *backupServer) AddFlags(cmd *cobra.Command) {
```
Comment (tjungblu) [inline pkg/cmd/backuprestore/backupserver.go:44]: why do we need to change to cobra here?

Comment (Elbehery) [reply pkg/cmd/backuprestore/backupserver.go:44]: so the `Prune` expects `*cobra.Command` , thats why

Comment (tjungblu) [reply pkg/cmd/backuprestore/backupserver.go:44]: then we can make that change when we introduce prune.Run

File: pkg/operator/periodicbackupcontroller/periodicbackupcontroller.go (hunk 20 above)
Comment (Elbehery) [reply pkg/operator/periodicbackupcontroller/periodicbackupcontroller.go:89]: gotcha 👍🏽

**Comment by Elbehery** [comment]
@tjungblu shall we label ?

**Comment by tjungblu** [comment]
you still haven't responded to me in:
https://github.com/openshift/cluster-etcd-operator/pull/1305/files#r1741523335
https://github.com/openshift/cluster-etcd-operator/pull/1305/files#r1741524122
https://github.com/openshift/cluster-etcd-operator/pull/1305/files#r1741524818

please don't resolve conversations that are not resolved yet

File: pkg/backuphelpers/backupvars.go (hunk 10 above)
Comment (Elbehery) [reply pkg/backuphelpers/backupvars.go:47]: ok, I am trying to avoid passing a ref around 

I make the copy within BackupVar then 👍🏽

Comment (tjungblu) [reply pkg/backuphelpers/backupvars.go:47]: Thank you.

File: pkg/cmd/backuprestore/backupserver.go (hunk 21 above)
Comment (Elbehery) [reply pkg/cmd/backuprestore/backupserver.go:44]: 👍🏽

File: pkg/backuphelpers/backupvars_test.go (hunk 17 above)
Comment (Elbehery) [reply pkg/backuphelpers/backupvars_test.go:31]: I actually removed the `bool` from the input, the code now handles this flag according to the spec being `nil`

File: pkg/backuphelpers/backupvars.go (hunk 10 above)
Comment (Elbehery) [reply pkg/backuphelpers/backupvars.go]: done

**Comment by Elbehery** [comment]
`TestNodeDownRevisionSaving` passes locally 

/retest-required

File: pkg/backuphelpers/backupvars.go [hunk 22]
```diff
@@ -0,0 +1,98 @@
...
+		return
+	}
+
+	b.enabled = spec != nil
+
+	if spec == nil {
+		b.spec = nil
+	} else {
+		specClone := *spec
+		b.spec = &specClone
```
Comment (tjungblu) [inline pkg/backuphelpers/backupvars.go:51]: I leave this for another PR, but you could also simply do a 
```suggestion
		b.spec = spec.DeepCopy()
```

File: pkg/backuphelpers/backupvars_test.go [hunk 23]
```diff
@@ -0,0 +1,113 @@
...
+		cr       *backupv1alpha1.EtcdBackupSpec
+		expected string
+	}{
+		{
+			"enabled",
+			createEtcdBackupSpec(timezone, schedule),
+			"    args:\n    - --enabled=true\n    - --timezone=GMT\n    - --schedule=0 */2 * * *",
+		},
+		{
+			"disabled",
```
Comment (tjungblu) [inline pkg/backuphelpers/backupvars_test.go:29]: disabled means "nil" now, maybe adjust the test to match this?

File: pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go [hunk 24]
```diff
@@ -88,14 +96,45 @@ func TestTargetConfigController(t *testing.T) {
...
-			eventRecorder, _, controller := getController(t, scenario.staticPodStatus, scenario.objects, scenario.etcdMembers)
+			backupVar := backuphelpers.NewDisabledBackupConfig()
+			eventRecorder, _, controller, fakeKubeClient := getController(t, scenario.staticPodStatus, scenario.objects, scenario.etcdMembers, backupVar)
+			if scenario.etcdBackupSpec != nil {
+				backupVar.SetBackupSpec(scenario.etcdBackupSpec)
+			}
 			err := controller.sync(context.TODO(), factory.NewSyncContext("test", eventRecorder))
 			assert.Equal(t, scenario.expectedErr, err)
+
+			if scenario.etcdBackupSpec != nil {
```
Comment (tjungblu) [inline pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go:132]: if it's nil, we should also assert the "--enabled=false" path here

File: pkg/operator/periodicbackupcontroller/periodicbackupcontroller_test.go [hunk 25]
```diff
@@ -114,6 +116,7 @@ func TestSyncLoopFailsDegradesOperator(t *testing.T) {
 		backupsClient:         operatorFake.ConfigV1alpha1(),
 		kubeClient:            client,
 		operatorImagePullSpec: "pullspec-image",
+		backupVarGetter:       backuphelpers.NewDisabledBackupConfig(),
```
Comment (tjungblu) [inline pkg/operator/periodicbackupcontroller/periodicbackupcontroller_test.go:119]: I think we also need to add a controller level test here that shows you:
* by default have no backup set ("NewDisabledBackupConfig")
* when there is a default CRD, you set it in the backupVarGetter
* when you remove the CRD again it's back to "no backup set" like the first case

File: pkg/backuphelpers/backupvars_test.go (hunk 23 above)
Comment (Elbehery) [reply pkg/backuphelpers/backupvars_test.go:29]: 👍🏽

File: pkg/backuphelpers/backupvars_test.go [hunk 26]
```diff
@@ -0,0 +1,113 @@
...
+
+			require.Equal(t, tc.expected, act)
+		})
+	}
+}
+
+func createEtcdBackupSpec(timezone, schedule string) *backupv1alpha1.EtcdBackupSpec {
+	if timezone == "" && schedule == "" {
+		return nil
+	}
```
Comment (tjungblu) [inline pkg/backuphelpers/backupvars_test.go:92]: ah okay that's how I got confused. Can you please remove this and just explicitly set "nil" in the disabled cases above?

File: pkg/backuphelpers/backupvars_test.go [hunk 27]
```diff
@@ -0,0 +1,113 @@
...
+		{
+			"empty timezone and schedule disabled",
+			createEtcdBackupSpec("", ""),
+			"    args:\n    - --enabled=false",
+		},
+		{
+			"empty timezone and schedule enabled",
+			createEtcdBackupSpec("", ""),
+			"    args:\n    - --enabled=false",
+		},
```
Comment (tjungblu) [inline pkg/backuphelpers/backupvars_test.go:63]: that's the exact same test

File: pkg/testutils/testutils.go [hunk 28]
```diff
@@ -540,3 +541,15 @@ func AssertNodeSpecificCertificates(t *testing.T, node *corev1.Node, secrets []c
 
 	require.Equalf(t, 0, expectedSet.Len(), "missing certificates for node: %v", expectedSet.List())
 }
+
+func CreateEtcdBackupSpec(timezone, schedule string) backupv1alpha1.EtcdBackupSpec {
```
Comment (tjungblu) [inline pkg/testutils/testutils.go:545]: if it's still unused, just return a pointer to the spec directly, no need for the other function

File: pkg/backuphelpers/backupvars.go (hunk 22 above)
Comment (Elbehery) [reply pkg/backuphelpers/backupvars.go:51]: done 👍🏽

File: pkg/backuphelpers/backupvars_test.go (hunk 23 above)
Comment (Elbehery) [reply pkg/backuphelpers/backupvars_test.go:29]: done 👍🏽

File: pkg/backuphelpers/backupvars_test.go (hunk 26 above)
Comment (Elbehery) [reply pkg/backuphelpers/backupvars_test.go:92]: but i can not create a Spec without `schedule`, or ?

File: pkg/backuphelpers/backupvars_test.go (hunk 27 above)
Comment (Elbehery) [reply pkg/backuphelpers/backupvars_test.go:63]: yes i cleaned those all 👍🏽

File: pkg/backuphelpers/backupvars_test.go (hunk 26 above)
Comment (Elbehery) [reply pkg/backuphelpers/backupvars_test.go:92]: done, but this way we can have a `BackupSpec` without `schedule`, i think this is not correct !!

Comment (tjungblu) [reply pkg/backuphelpers/backupvars_test.go:92]: that's the purpose of a unit test, isn't it? ;)

File: pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go (hunk 24 above)
Comment (Elbehery) [reply pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go:157]: added, however this broke another case (i.e. `Quorum not tolerant`, since there was `CM[etcdpod.yaml]` to assert

Therefore this assertion only takes place iff the scenario has no error

File: pkg/backuphelpers/backupvars_test.go (hunk 26 above)
Comment (Elbehery) [reply pkg/backuphelpers/backupvars_test.go]: :D :D .. yes i know, but i think we should `not` enable the container if the schedule is empty, or ?

File: pkg/testutils/testutils.go (hunk 28 above)
Comment (Elbehery) [reply pkg/testutils/testutils.go]: done 👍🏽

File: pkg/backuphelpers/backupvars_test.go (hunk 26 above)
Comment (tjungblu) [reply pkg/backuphelpers/backupvars_test.go]: you can't create such a CR because it's a required field in the apiserver

File: pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go (hunk 24 above)
Comment (tjungblu) [reply pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go:157]: yes, because in that case the controller has no output. Can we only run the assertions if the "etcdpod.yaml" exists in the configmap instead?

Comment (tjungblu) [reply pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go:157]: I'll send a suggestion, one sec, had a caching issue

File: pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go [hunk 29]
```diff
@@ -88,14 +96,70 @@ func TestTargetConfigController(t *testing.T) {
...
+			etcdPodCM, err := fakeKubeClient.CoreV1().ConfigMaps(operatorclient.TargetNamespace).Get(context.TODO(), "etcd-pod", metav1.GetOptions{})
+			require.NoError(t, err)
+			var expStr string
+
+			if scenario.etcdBackupSpec != nil {
+				expStr = "    args:\n    - --enabled=true\n    - --timezone=GMT\n    - --schedule=0 */2 * * *"
+			} else {
+				expStr = "    args:\n    - --enabled=false"
+			}
+			require.Contains(t, etcdPodCM.Data["pod.yaml"], expStr)
```
Comment (tjungblu) [inline pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go:162]:
```suggestion

    expStr := " args:\n - --enabled=false"
    if scenario.etcdBackupSpec != nil {
       expStr = " args:\n - --enabled=true\n - --timezone=GMT\n - --schedule=0 */2 * * *"
    }
    

   if _, ok := etcdPodCM.Data["pod.yaml"]; ok {
       require.Contains(t, etcdPodCM.Data["pod.yaml"], expStr)
   }
			
```

File: pkg/backuphelpers/backupvars_test.go (hunk 26 above)
Comment (Elbehery) [reply pkg/backuphelpers/backupvars_test.go]: ahhh .. so it will be rejected by the APIServer  .. alright thats a relief

File: pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go (hunk 29 above)
Comment (Elbehery) [reply pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go:162]: lgtm 👍🏽

File: pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go [hunk 30]
```diff
@@ -88,14 +96,70 @@ func TestTargetConfigController(t *testing.T) {
...
+				u.WithNodeStatusAtCurrentRevision(3),
+				u.WithNodeStatusAtCurrentRevision(3),
+			),
+			etcdMembers: []*etcdserverpb.Member{
+				u.FakeEtcdMemberWithoutServer(0),
+				u.FakeEtcdMemberWithoutServer(1),
+				u.FakeEtcdMemberWithoutServer(2),
+			},
+			etcdBackupSpec:    u.CreateEtcdBackupSpecPtr("GMT", "0 */2 * * *"),
+			etcdMembersEnvVar: "1,2,3",
```
Comment (tjungblu) [inline pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go:119]: I've just noticed and this is not related to your PR, but is this envvar even used?

File: pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go (hunk 29 above)
Comment (tjungblu) [reply pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go:162]: sorry, the expectedErr early return also does the same job

```
			if scenario.expectedErr != nil {
				return
			}
```

File: pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go (hunk 30 above)
Comment (Elbehery) [reply pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go:119]: not at all, shall i remove ?

File: pkg/backuphelpers/backupvars_test.go (hunk 26 above)
Comment (tjungblu) [reply pkg/backuphelpers/backupvars_test.go]: apparently it's not, but it's certainly something we should add for a new version of the CRD then:
https://github.com/openshift/api/blob/master/config/v1alpha1/types_backup.go#L51

Then again "Empty string means no opinion and the platform is left to choose a reasonable default which is subject to change without notice." feel free to set something by default in the next PR in the sidecar logic

File: pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go (hunk 30 above)
Comment (tjungblu) [reply pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go:119]: I'll do another PR after this one merged, let's not further complicate :rofl:

File: pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go (hunk 29 above)
Comment (Elbehery) [reply pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go:162]: mir egal :D

Comment (tjungblu) [reply pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go:162]: mir auch, let's leave it the way it is

Comment (Elbehery) [reply pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go:162]: 👍🏽

File: pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go (hunk 30 above)
Comment (Elbehery) [reply pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go:119]: sgtm 👍🏽

File: pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go (hunk 24 above)
Comment (Elbehery) [reply pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go:157]: done 👍🏽

File: pkg/backuphelpers/backupvars_test.go (hunk 26 above)
Comment (Elbehery) [reply pkg/backuphelpers/backupvars_test.go]: lgtm 👍🏽

File: pkg/operator/periodicbackupcontroller/periodicbackupcontroller_test.go (hunk 25 above)
Comment (Elbehery) [reply pkg/operator/periodicbackupcontroller/periodicbackupcontroller_test.go:183]: created this test https://github.com/openshift/cluster-etcd-operator/pull/1305/commits/f21b98a8afc441a3b9ebb8a14436c9c76c697124

**Comment by tjungblu** [comment]
well done, thanks for your patience.

/lgtm
